core/domain/question_domain.py

```diff
@@ -215,6 +216,168 @@ def _convert_state_v28_dict_to_v29_dict(cls, question_state_dict):
         return question_state_dict
 
+    @classmethod
+    def _convert_state_v29_dict_to_v30_dict(cls, question_state_dict):
+        """Converts from version 29 to 30. Version 30 add image assets
+        in the question_state_dict. This funtion is extracting HTMl of state
+        and do 3 operations on each HTML content.
+        The 3 operations and their sequence were.
+        1) Add image_id in each image tag present in the HTML.
+        2) Extract image info of each image from HTML.
+        3) Remove filepath of each image present in the HTML.
+
+        Args:
+            question_state_dict: dict. The dict representation of
+                question_state_data.
+
+        Returns:
+            dict. The converted question_state_dict.
+        """
+        image_counter = 0
+        image_assets = {}
+        image_mapping = {}
+        image_info_list = []
+
+        # Adding image_id, getting image info and removing filepath from
+        # image tag present in state content.
+        content_html = question_state_dict['content']['html']
+        (new_content_html, image_counter) = (
+            html_validation_service.add_image_id_in_image_tag(
+                content_html, image_counter))
+
+        image_info_list = image_info_list + (
+            html_validation_service.get_image_info_from_html(
+                new_content_html))
+
+        question_state_dict['content']['html'] = (
+            html_validation_service.remove_filepath_from_image_tag(
+                new_content_html))
+
+        # Adding image_id, getting image info and removing filepath from
+        # image tag present in state interaction.
+        if question_state_dict['interaction']['default_outcome']:
+            interaction_feedback_html = question_state_dict[
+                'interaction']['default_outcome']['feedback']['html']
+            (new_interaction_feedback_html, image_counter) = (
+                html_validation_service.add_image_id_in_image_tag(
+                    interaction_feedback_html, image_counter))
+
+            image_info_list = image_info_list + (
+                html_validation_service.get_image_info_from_html(
+                    new_interaction_feedback_html))
+
+            (
+                question_state_dict['interaction']['default_outcome']
+                ['feedback']['html']) = (
+                    html_validation_service.
+                    remove_filepath_from_image_tag(
+                        new_interaction_feedback_html))
+
+        # Adding image_id, getting image info and removing filepath from
+        # image tag present in state interaction answer_groups.
+        for answer_group_index, answer_group in enumerate(
+                question_state_dict['interaction']['answer_groups']):
+            answer_group_html = answer_group['outcome']['feedback']['html']
+            (new_answer_group_html, image_counter) = (
+                html_validation_service.add_image_id_in_image_tag(
```

Comment:
Ditto. I'm leaving the review here as I'm expecting the backend tests to fail. Please do a self-review, take your time!